### PR TITLE
SI-8574 Copy @SerialVersionUID, etc, to specialized subclasses

### PR DESCRIPTION
**Pull request by retronym**
The test case demonstrates that this is important for serialization
and for strictfp. (Although the latter is still pretty broken,
see SI-7954.)

### CODE REVIEW

**Comment by retronym** [comment]
Review by @gkossakowski, @lrytz

**Comment by retronym** [comment]
Ah, I probably shouldn't copy down `@deprecated`.

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala [hunk 1]
```diff
@@ -538,6 +538,7 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
       bytecodeClazz.info
 
       val sClass = clazz.owner.newClass(clazzName, clazz.pos, (clazz.flags | SPECIALIZED) & ~CASE)
+      sClass.setAnnotations(clazz.annotations.filterNot(_.symbol == DeprecatedAttr))
```
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala]: Why deprecated is excluded? To not generate deprecation warning about synthetic classes?

**Comment by retronym** [comment]
@gkossakowski I found a better way to deal with the double deprecated-inheritance warnings, see the comment in the latest commit.

**Comment by retronym** [comment]
PLS REBUILD ALL

**Comment by retronym** [comment]
[pr-scala-integrate-partest](https://scala-webapps.epfl.ch/jenkins/job/pr-scala-integrate-partest/5762/console)

```
/localhome/jenkins/c/workspace/pr-scala-integrate-partest/sbt-launch.jar 'reboot full' clean 'set every scalaVersion := "2.11.1-39a64df-SNAPSHOT"' 'show scala-instance' 'show scalaVersion' 'show scalaBinaryVersion' publish-local
Invalid or corrupt jarfile /localhome/jenkins/c/workspace/pr-scala-integrate-partest/sbt-launch.jar
Build step 'Execute shell' marked build as failure
Notifying upstream projects of job completion
Finished: FAILURE
```

**Comment by adriaanm** [comment]
restarted ide build (only failure): https://scala-webapps.epfl.ch/jenkins/job/pr-scala-integrate-ide/7045/console

**Comment by retronym** [comment]
PLS REBUILD ALL

**Comment by retronym** [comment]
ping @gkossakowski @lrytz for review for 2.11.1

File: test/files/run/t8574.scala [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  }
+  def checkStrictFp(cls: Class[_]) = {
+    import java.lang.reflect._
+    for (m <- cls.getDeclaredMethods) {
+      val isStrict = Modifier.isStrict(m.getModifiers)
+      assert(isStrict, cls)  
+    }
+  }
+  def check(x: AnyRef) {
+    // checkUID(x.getClass, 42)
```
Comment (lrytz) [inline test/files/run/t8574.scala]: should that be un-commented?

**Comment by lrytz** [comment]
LGTM for specialization. for the general case we could have a meta-annotation to control wether an annotation is copied to subclasses.

**Comment by retronym** [comment]
I just uncommented that part of the test and force pushed. I had commented it out locally to make sure that both parts failed without the change to the compiler.

**Comment by retronym** [comment]
Outside of `@strictfp` and `@SerialVersionUID`, can we think of other annotations that influence the compiler post-specialization?

**Comment by retronym** [comment]
```
val isNative         = methSymbol.hasAnnotation(definitions.NativeAttr)
```

... for methods, not classes

**Comment by retronym** [comment]
Here's one I've never noticed before:

```
/**
 * An annotation that designates the class to which it is applied as remotable.
 *
 * For instance, the Scala code
 * {{{
 * @remote trait Hello {
 *   def sayHello(): String
 * }
 * }}}
 * is equivalent to the following Java code:
 * {{{
 * public interface Hello extends java.rmi.Remote {
 *     String sayHello() throws java.rmi.RemoteException;
 * }
 * }}}
 */
class remote extends scala.annotation.StaticAnnotation {}
```

**Comment by retronym** [comment]
Fun comment in the backend:

``` scala
/* Adds a @remote annotation, actual use unknown.
     *
     * Invoked from genMethod() and addForwarder().
     *
     * must-single-thread
     */
```

**Comment by retronym** [comment]
Next: `@transient` and `@volatile`, also not relevant for classes.

**Comment by adriaanm** [comment]
`@remote` was probably part of micheloud's thesis, I don't think there's any code in the compiler that relies on it

**Pull request merged**